### PR TITLE
prov/mrail: Add message ordering

### PR DESCRIPTION
**Pull request by yburette**

### CODE REVIEW

File: prov/mrail/src/mrail_cq.c [hunk 1]
```diff
@@ -163,6 +142,140 @@ static int mrail_cq_process_comp_buf_recv(struct util_cq *cq,
...
+{
+	struct slist *queue = &peer_addr->early_recv_comps;
+	struct mrail_early_comp *early_comp;
+
+	if (!slist_empty(queue)) {
+		early_comp = container_of(queue->head, struct
+				mrail_early_comp, entry);
+		if (early_comp->seq_no == peer_addr->expected_seq_no) {
+			slist_remove_head(queue);
+			peer_addr->expected_seq_no++;
```
Comment (shefty) [inline prov/mrail/src/mrail_cq.c]: I don't think the locking is correct for this patch.  We probably don't need the locks in mrail_alloc_early_comp() and mrail_free_early_comp() -- because we should likely already be holding one.  There needs to be locks around accessing the expected_seq_no value, and the early_recv_comp queue.

Btw, I would rename early_recv_comp to something like ooo_recv_queue, or at least something with the word queue in the name.

**Comment by rajachan** [comment]
Why not use the `ofi_recvwin.h` utility instead of an ordered slist? You get constant-order insertions with the rewcvwin utility.

**Comment by shefty** [comment]
Mrail assumes a reliable channel underneath, so any retransmissions or windowing would be the responsibility of the underlying rail EP.  So I don't think recvwin matches up as well in this case.

File: prov/mrail/src/mrail_cq.c [hunk 2]
```diff
@@ -111,22 +113,133 @@ int mrail_cq_process_buf_recv(mrail_cq_entry_t *comp, struct mrail_recv *recv)
...
+	}
+	mrail_ep->lock_release(&mrail_ep->util_ep.lock);
+	if (OFI_UNLIKELY(!recv))
+		return 0;
+
+	return mrail_cq_process_buf_recv(comp, recv);
+}
+
+static
+struct mrail_early_comp *mrail_peer_get_next(struct mrail_peer_info *peer_info)
```
Comment (a-ilango) [inline prov/mrail/src/mrail_cq.c]: nit: rename this to mrail_get_early_comp

**Comment by yburette** [comment]
I think this covers everything that we discussed. Please let me know if I have missed anything.

File: prov/mrail/src/mrail.h [hunk 3]
```diff
@@ -212,9 +212,6 @@ struct mrail_ep {
 	struct util_buf_pool	*req_pool;
 	struct util_buf_pool 	*ooo_recv_pool;
 	struct slist		deferred_reqs;
-
-	ofi_fastlock_acquire_t	lock_acquire;
-	ofi_fastlock_release_t	lock_release;
```
Comment (shefty) [inline prov/mrail/src/mrail.h]: The commit message doesn't describe why this change is needed.  Please elaborate

File: include/ofi_util.h [hunk 4]
```diff
@@ -256,6 +256,8 @@ struct util_ep {
 	ofi_ep_progress_func	progress;
 	struct util_cmap	*cmap;
 	fastlock_t		lock;
+	ofi_fastlock_acquire_t	ep_lock_acquire;
+	ofi_fastlock_release_t	ep_lock_release;
```
Comment (shefty) [inline include/ofi_util.h]: I'm reading the commits backwards -- please merge this commit with the next one.

File: prov/mrail/src/mrail_cq.c [hunk 5]
```diff
@@ -164,17 +161,32 @@ static int mrail_process_ooo_recvs(struct mrail_ep *mrail_ep,
...
+	}
+	while (ooo_recv) {
 		FI_DBG(&mrail_prov, FI_LOG_CQ, "found ooo_recv seq=%d\n",
 				ooo_recv->seq_no);
-		ret = mrail_cq_process_comp_buf_recv(mrail_ep, &ooo_recv->comp);
-		if (ret)
-			return ret;
-		mrail_free_ooo_recv(mrail_ep, ooo_recv);
-	}
+		recv = mrail_match_recv(mrail_ep, &ooo_recv->comp);
```
Comment (shefty) [inline prov/mrail/src/mrail_cq.c]: Why is it safe to drop the lock while still accessing oob_recv and recv?  Can another thread acquire the same objects and modify either?  Does mrail_match_recv remove recv so that only this thread has access to it?

Comment (shefty) [reply prov/mrail/src/mrail_cq.c]: This is the information I was looking for.  I was just asking for clarification, so I didn't have to read all of the code.  :)

File: prov/mrail/src/mrail_cq.c [hunk 6]
```diff
@@ -113,22 +113,145 @@ int mrail_cq_process_buf_recv(struct fi_cq_tagged_entry *comp,
 	return retv;
 }
 
-static int mrail_cq_process_comp_buf_recv(struct util_cq *cq,
-					  struct fi_cq_tagged_entry *comp)
+/* Should only be called while holding a lock */
```
Comment (shefty) [inline prov/mrail/src/mrail_cq.c]: need to specify what lock must be held

File: prov/mrail/src/mrail_cq.c [hunk 7]
```diff
@@ -113,22 +113,145 @@ int mrail_cq_process_buf_recv(struct fi_cq_tagged_entry *comp,
...
+{
+	struct mrail_ooo_recv *ooo_recv;
+	struct mrail_recv *recv;
+	int ret;
+
+	ofi_ep_lock_acquire(&mrail_ep->util_ep);
+	ooo_recv = mrail_get_next_recv(peer_info);
+	if (!ooo_recv) {
+		goto exit;
+	}
```
Comment (shefty) [inline prov/mrail/src/mrail_cq.c]: This if statement can be dropped.  The while statement does the same check and ends at the same location.

File: prov/mrail/src/mrail.h [hunk 8]
```diff
@@ -172,13 +172,21 @@ struct mrail_av {
 	ofi_atomic32_t index;
 };
 
-struct mrail_peer_addr {
+struct mrail_peer_info {
+	struct slist	ooo_recv_queue;
 	fi_addr_t	addr;
 	uint32_t	seq_no;
+	uint32_t	expected_seq_no;
```
Comment (shefty) [inline prov/mrail/src/mrail.h]: This patch should have actually been at least 3.  1. rename mrail_peer_addr to mrail_peer_info.  2. replace mrail ep lock with util ep lock.  3. handle out of order messages.

Maybe I wasn't clear in my previous message.  What a series should not contain is a patch that updates or fixes an earlier patch in the series.  Don't respin this patch to break it back apart.  Just update based on the other comments.

Comment (yburette) [reply prov/mrail/src/mrail.h]: Understood. I'll update this with your comments.

**Comment by shefty** [comment]
Is this PR good to merge?

**Comment by yburette** [comment]
I believe so, yes.
